e2e/testcases/hydration_test.go

```diff
@@ -445,3 +472,262 @@ func TestHydrateResourcesInRelativePath(t *testing.T) {
+	for _, errCode := range errCodes {
+		errorList = append(errorList, v1beta1.ConfigSyncError{
+			Code: errCode,
+			// ErrorMessage ignored
+			// Resources ignored
+		})
+	}
+
+	// Validate .status.rendering fields.
+	expectedRenderingStatus := v1beta1.RenderingStatus{
+		Git: &v1beta1.GitStatus{
+			Repo:     rs.Spec.Repo,
+			Revision: gitRevisionOrDefault(*rs.Spec.Git),
+			Branch:   rs.Spec.Git.Branch,
+			Dir:      rs.Spec.Git.Dir,
+		},
+		// LastUpdate ignored
+		Message:      parse.RenderingFailed,
+		Commit:       commit,
+		Errors:       errorList,
+		ErrorSummary: errorSummary,
+	}
+	assertEqual(nt, asserter, expectedRenderingStatus, rs.Status.Rendering,
+		"RootSync .status.rendering")
+
+	// Validate Syncing condition fields
+	rsSyncingCondition := rootsync.GetCondition(rs.Status.Conditions, v1beta1.RootSyncSyncing)
+	expectedSyncingCondition := &v1beta1.RootSyncCondition{
+		Type: v1beta1.RootSyncSyncing,
+		// Status ignored
+		// LastUpdateTime ignored
+		// LastTransitionTime ignored
+		Reason:  "Rendering",
+		Message: "Rendering failed",
+		Commit:  commit,
+		// Errors unused by the Syncing condition (always nil)
+		ErrorSourceRefs: errorSources,
+		ErrorSummary:    errorSummary,
+	}
+	assertEqual(nt, asserter, expectedSyncingCondition, rsSyncingCondition,
+		"RootSync .status.conditions[.status=%q]", v1beta1.RootSyncSyncing)
+
+	if nt.T.Failed() {
+		nt.T.FailNow()
+	}
+}
+
+func validateRootSyncSourceErrors(nt *nomostest.NT, rs *v1beta1.RootSync, commit string, errCodes ...string) {
+	nt.T.Helper()
+
+	if len(errCodes) == 0 {
+		nt.T.Fatal("Invalid test: expected specific errors to validate, but none were specified")
+	}
+
+	// Use a custom asserter so we can ignore hard-to-test fields.
+	// Testing whole structs makes debugging easier by printing the full
+	// expected and actual values, but it will also print any ignored fields.
+	asserter := testutil.NewAsserter(
+		cmpopts.IgnoreFields(v1beta1.SourceStatus{}, "LastUpdate"),
+		cmpopts.IgnoreFields(v1beta1.RootSyncCondition{}, "LastUpdateTime", "LastTransitionTime", "Message"),
+		cmpopts.IgnoreFields(v1beta1.ConfigSyncError{}, "ErrorMessage", "Resources"),
+		// Ignore the current Syncing condition status. Retries will cause flapping between True & False.
+		cmpopts.IgnoreFields(v1beta1.RootSyncCondition{}, "Status"),
+	)
+
+	// Build untruncated ErrorSummary & fake Errors list from error codes
+	var errorList []v1beta1.ConfigSyncError
+	var errorSources []v1beta1.ErrorSource
+
+	errorSources = append(errorSources, v1beta1.SourceError)
+	errorSummary := &v1beta1.ErrorSummary{
+		TotalCount:                len(errCodes),
+		Truncated:                 false,
+		ErrorCountAfterTruncation: len(errCodes),
+	}
+	for _, errCode := range errCodes {
+		errorList = append(errorList, v1beta1.ConfigSyncError{
+			Code: errCode,
+			// ErrorMessage ignored
+			// Resources ignored
+		})
+	}
+
+	// Validate .status.source fields.
+	expectedRenderingStatus := v1beta1.SourceStatus{
```

Comment:
Confirming it only checks for source error in the rendering status? Might worth clarify that in the func name

Reply:
renamed expectedSourceStatus